File: src/protocol.zig
const std = @import("std");

pub const ready_marker = "\x1b]133;A\x07";
pub const prompt_end_marker = "\x1b]133;B\x07";
pub const command_begin_marker = "\x1b]133;C\x07";
pub const command_done_prefix = "\x1b]133;D;";

pub const Request = struct {
    id: []u8,
    command: []u8,
};

pub const Completion = struct {
    id: []u8,
    exit_code: i32,
    timed_out: bool,
};

pub fn buildExecCommand(allocator: std.mem.Allocator, command: []const u8) ![]u8 {
    if (command.len != 0 and command[command.len - 1] == '\n') {
        return allocator.dupe(u8, command);
    }
    return std.fmt.allocPrint(allocator, "{s}\n", .{command});
}

pub fn stringifyRequest(allocator: std.mem.Allocator, request: Request) ![]u8 {
    return std.json.stringifyAlloc(allocator, request, .{ .whitespace = .indent_2 });
}

pub fn parseRequest(allocator: std.mem.Allocator, bytes: []const u8) !Request {
    const parsed = try std.json.parseFromSlice(Request, allocator, bytes, .{ .ignore_unknown_fields = true });
    defer parsed.deinit();
    const id = try allocator.dupe(u8, parsed.value.id);
    errdefer allocator.free(id);
    return .{
        .id = id,
        .command = try allocator.dupe(u8, parsed.value.command),
    };
}

pub fn stringifyCompletion(allocator: std.mem.Allocator, completion: Completion) ![]u8 {
    return std.json.stringifyAlloc(allocator, completion, .{ .whitespace = .indent_2 });
}

pub fn parseCompletion(allocator: std.mem.Allocator, bytes: []const u8) !Completion {
    const parsed = try std.json.parseFromSlice(Completion, allocator, bytes, .{ .ignore_unknown_fields = true });
    defer parsed.deinit();
    const id = try allocator.dupe(u8, parsed.value.id);
    return .{
        .id = id,
        .exit_code = parsed.value.exit_code,
        .timed_out = parsed.value.timed_out,
    };
}

pub const StdoutParse = struct {
    wrote_data: bool,
    began_request: bool,
    ended_request: bool,
    finished_prompt: bool,
    exit_code: ?i32,
};

pub const StdoutParseState = struct {
    inside_prompt: bool = false,
    pending: [256]u8 = undefined,
    pending_len: usize = 0,
};

pub fn parseStdoutChunk(writer: anytype, chunk: []const u8, state: *StdoutParseState) !StdoutParse {
    if (state.pending_len != 0) {
        const pending_len = state.pending_len;
        const merged_len = pending_len + chunk.len;
        if (merged_len <= 8192) {
            var merged: [8192]u8 = undefined;
            @memcpy(merged[0..pending_len], state.pending[0..pending_len]);
            @memcpy(merged[pending_len..merged_len], chunk);
            state.pending_len = 0;
            return parseStdoutChunkNoPending(writer, merged[0..merged_len], state);
        }
        try writer.writeAll(state.pending[0..pending_len]);
        state.pending_len = 0;
    }

    return parseStdoutChunkNoPending(writer, chunk, state);
}

fn parseStdoutChunkNoPending(writer: anytype, chunk: []const u8, state: *StdoutParseState) !StdoutParse {
    var result = StdoutParse{
        .wrote_data = false,
        .began_request = false,
        .ended_request = false,
        .finished_prompt = false,
        .exit_code = null,
    };

    var index: usize = 0;
    while (index < chunk.len) {
        const escape_index = std.mem.indexOfScalarPos(u8, chunk, index, 0x1b) orelse {
            if (!state.inside_prompt and chunk[index..].len != 0) {
                try writer.writeAll(chunk[index..]);
                result.wrote_data = true;
            }
            break;
        };

        if (!state.inside_prompt and escape_index > index) {
            try writer.writeAll(chunk[index..escape_index]);
            result.wrote_data = true;
        }

        const marker_end = std.mem.indexOfScalarPos(u8, chunk, escape_index, 0x07) orelse {
            const tail = chunk[escape_index..];
            if (tail.len <= state.pending.len) {
                @memcpy(state.pending[0..tail.len], tail);
                state.pending_len = tail.len;
            } else {
                try writer.writeAll(tail);
                result.wrote_data = true;
            }
            break;
        };

        const marker = chunk[escape_index .. marker_end + 1];
        if (std.mem.eql(u8, marker, ready_marker)) {
            state.inside_prompt = true;
        } else if (std.mem.eql(u8, marker, prompt_end_marker)) {
            state.inside_prompt = false;
            result.finished_prompt = true;
        } else if (std.mem.eql(u8, marker, command_begin_marker)) {
            state.inside_prompt = false;
            result.began_request = true;
        } else if (parseExitCode(marker)) |exit_code| {
            state.inside_prompt = false;
            result.ended_request = true;
            result.exit_code = exit_code;
        } else if (!state.inside_prompt) {
            try writer.writeAll(marker);
            result.wrote_data = true;
        }

        index = marker_end + 1;
    }

    return result;
}

pub fn flushStdoutChunk(writer: anytype, state: *StdoutParseState) !void {
    if (state.pending_len == 0) return;
    try writer.writeAll(state.pending[0..state.pending_len]);
    state.pending_len = 0;
}

pub fn chunkContainsReady(chunk: []const u8) bool {
    return std.mem.indexOf(u8, chunk, ready_marker) != null;
}

fn parseExitCode(marker: []const u8) ?i32 {
    if (!std.mem.startsWith(u8, marker, command_done_prefix)) return null;
    const body = marker[command_done_prefix.len .. marker.len - 1];
    return std.fmt.parseInt(i32, body, 10) catch null;
}

test "parse stdout strips generic markers" {
    var out = std.ArrayList(u8).init(std.testing.allocator);
    defer out.deinit();
    const chunk = command_begin_marker ++ "hello" ++ command_done_prefix ++ "3\x07";
    var state = StdoutParseState{};
    const parsed = try parseStdoutChunk(out.writer(), chunk, &state);
    try std.testing.expect(parsed.began_request);
    try std.testing.expect(parsed.ended_request);
    try std.testing.expect(!parsed.finished_prompt);
    try std.testing.expectEqual(@as(?i32, 3), parsed.exit_code);
    try std.testing.expectEqualStrings("hello", out.items);
}

test "parse stdout strips prompt text" {
    var out = std.ArrayList(u8).init(std.testing.allocator);
    defer out.deinit();
    var state = StdoutParseState{};
    const chunk = ready_marker ++ "user=> " ++ prompt_end_marker ++ command_begin_marker ++ "1\n" ++ command_done_prefix ++ "0\x07";
    const parsed = try parseStdoutChunk(out.writer(), chunk, &state);
    try std.testing.expect(parsed.began_request);
    try std.testing.expect(parsed.ended_request);
    try std.testing.expect(parsed.finished_prompt);
    try std.testing.expectEqualStrings("1\n", out.items);
}

test "parse stdout keeps split markers across chunks" {
    var out = std.ArrayList(u8).init(std.testing.allocator);
    defer out.deinit();
    var state = StdoutParseState{};
    const chunk1 = "hello" ++ command_begin_marker[0 .. command_begin_marker.len - 1];
    const parsed1 = try parseStdoutChunk(out.writer(), chunk1, &state);
    try std.testing.expect(!parsed1.began_request);
    try std.testing.expectEqualStrings("hello", out.items);

    const chunk2 = "\x07world" ++ command_done_prefix ++ "0\x07";
    const parsed2 = try parseStdoutChunk(out.writer(), chunk2, &state);
    try std.testing.expect(parsed2.began_request);
    try std.testing.expect(parsed2.ended_request);
    try std.testing.expectEqualStrings("helloworld", out.items);
}

test "ready accepts prompt marker" {
    try std.testing.expect(chunkContainsReady(ready_marker));
}

File: src/protocol.rs
use crate::{AppError, Result};
use std::io::Write;

pub const READY_MARKER: &str = "\x1b]133;A\x07";
pub const PROMPT_END_MARKER: &str = "\x1b]133;B\x07";
pub const COMMAND_BEGIN_MARKER: &str = "\x1b]133;C\x07";
pub const COMMAND_DONE_PREFIX: &str = "\x1b]133;D;";

#[derive(Debug, Clone)]
pub struct ExecRequest {
    pub reply_path: String,
    pub timeout_ms: u64,
    pub command: String,
}

#[derive(Debug, Clone)]
pub struct StopRequest {
    pub reply_path: String,
}

#[derive(Debug, Clone)]
pub struct Completion {
    pub exit_code: i32,
    pub timed_out: bool,
}

#[derive(Debug, Clone)]
pub enum FrameKind {
    Exec,
    Stop,
    Ready,
    Stdout,
    Stderr,
    Complete,
    Stopped,
    Err,
}

#[derive(Debug, Clone)]
pub struct Frame {
    pub kind: FrameKind,
    pub payload: Vec<u8>,
}

pub fn build_exec_command(command: &str) -> String {
    if command.ends_with('\n') {
        command.to_string()
    } else {
        let mut value = String::from(command);
        value.push('\n');
        value
    }
}

pub fn encode_exec_request(request: &ExecRequest) -> Vec<u8> {
    format!(
        "reply_path={}\ntimeout_ms={}\ncommand={}\n",
        request.reply_path,
        request.timeout_ms,
        hex_encode(request.command.as_bytes())
    )
    .into_bytes()
}

pub fn encode_stop_request(request: &StopRequest) -> Vec<u8> {
    format!("reply_path={}\n", request.reply_path).into_bytes()
}

pub fn parse_exec_request(bytes: &[u8]) -> Result<ExecRequest> {
    let mut reply_port = None;
    let mut timeout_ms = None;
    let mut command = None;
    let text =
        std::str::from_utf8(bytes).map_err(|_| AppError::Msg("invalid request".to_string()))?;
    for line in text.lines() {
        let (key, value) = line
            .split_once('=')
            .ok_or_else(|| AppError::Msg("invalid request".to_string()))?;
        match key {
            "reply_path" => {
                reply_port = Some(value.to_string());
            }
            "timeout_ms" => {
                timeout_ms = Some(
                    value
                        .parse::<u64>()
                        .map_err(|_| AppError::Msg("invalid request".to_string()))?,
                )
            }
            "command" => {
                let bytes = hex_decode(value)?;
                command = Some(
                    String::from_utf8(bytes)
                        .map_err(|_| AppError::Msg("invalid request".to_string()))?,
                )
            }
            _ => {}
        }
    }
    Ok(ExecRequest {
        reply_path: reply_port.ok_or_else(|| AppError::Msg("invalid request".to_string()))?,
        timeout_ms: timeout_ms.ok_or_else(|| AppError::Msg("invalid request".to_string()))?,
        command: command.ok_or_else(|| AppError::Msg("invalid request".to_string()))?,
    })
}

pub fn parse_stop_request(bytes: &[u8]) -> Result<StopRequest> {
    let mut reply_path = None;
    let text =
        std::str::from_utf8(bytes).map_err(|_| AppError::Msg("invalid request".to_string()))?;
    for line in text.lines() {
        let (key, value) = line
            .split_once('=')
            .ok_or_else(|| AppError::Msg("invalid request".to_string()))?;
        if key == "reply_path" {
            reply_path = Some(value.to_string());
        }
    }
    Ok(StopRequest {
        reply_path: reply_path.ok_or_else(|| AppError::Msg("invalid request".to_string()))?,
    })
}

pub fn encode_completion(completion: &Completion) -> Vec<u8> {
    let mut payload = Vec::with_capacity(5);
    payload.extend_from_slice(&completion.exit_code.to_le_bytes());
    payload.push(u8::from(completion.timed_out));
    payload
}

pub fn decode_completion(payload: &[u8]) -> Result<Completion> {
    if payload.len() != 5 {
        return Err(AppError::Msg("invalid completion payload".to_string()));
    }
    let mut code = [0u8; 4];
    code.copy_from_slice(&payload[..4]);
    Ok(Completion {
        exit_code: i32::from_le_bytes(code),
        timed_out: payload[4] != 0,
    })
}

pub fn chunk_contains_ready(chunk: &[u8]) -> bool {
    chunk
        .windows(READY_MARKER.len())
        .any(|window| window == READY_MARKER.as_bytes())
}

#[derive(Debug, Clone)]
pub struct StdoutParseState {
    pub inside_prompt: bool,
    pub pending: Vec<u8>,
}

#[derive(Debug, Clone)]
pub struct StdoutParse {
    pub wrote_data: bool,
    pub began_request: bool,
    pub ended_request: bool,
    pub finished_prompt: bool,
    pub exit_code: Option<i32>,
}

pub fn parse_stdout_chunk(
    mut writer: impl Write,
    chunk: &[u8],
    state: &mut StdoutParseState,
) -> std::io::Result<StdoutParse> {
    if !state.pending.is_empty() {
        let mut merged = state.pending.clone();
        merged.extend_from_slice(chunk);
        state.pending.clear();
        return parse_stdout_chunk_no_pending(&mut writer, &merged, state);
    }
    parse_stdout_chunk_no_pending(&mut writer, chunk, state)
}

fn parse_stdout_chunk_no_pending(
    writer: &mut impl Write,
    chunk: &[u8],
    state: &mut StdoutParseState,
) -> std::io::Result<StdoutParse> {
    let mut result = StdoutParse {
        wrote_data: false,
        began_request: false,
        ended_request: false,
        finished_prompt: false,
        exit_code: None,
    };
    let mut index = 0;
    while index < chunk.len() {
        let escape = match chunk[index..].iter().position(|b| *b == 0x1b) {
            Some(offset) => index + offset,
            None => {
                if !state.inside_prompt && index < chunk.len() {
                    writer.write_all(&chunk[index..])?;
                    result.wrote_data = true;
                }
                break;
            }
        };
        if !state.inside_prompt && escape > index {
            writer.write_all(&chunk[index..escape])?;
            result.wrote_data = true;
        }
        let marker_end = match chunk[escape..].iter().position(|b| *b == 0x07) {
            Some(offset) => escape + offset,
            None => {
                state.pending.extend_from_slice(&chunk[escape..]);
                break;
            }
        };
        let marker = &chunk[escape..=marker_end];
        if marker == READY_MARKER.as_bytes() {
            state.inside_prompt = true;
        } else if marker == PROMPT_END_MARKER.as_bytes() {
            state.inside_prompt = false;
            result.finished_prompt = true;
        } else if marker == COMMAND_BEGIN_MARKER.as_bytes() {
            state.inside_prompt = false;
            result.began_request = true;
        } else if let Some(exit_code) = parse_exit_code(marker) {
            state.inside_prompt = false;
            result.ended_request = true;
            result.exit_code = Some(exit_code);
        } else if !state.inside_prompt {
            writer.write_all(marker)?;
            result.wrote_data = true;
        }
        index = marker_end + 1;
    }
    Ok(result)
}

fn parse_exit_code(marker: &[u8]) -> Option<i32> {
    if !marker.starts_with(COMMAND_DONE_PREFIX.as_bytes()) {
        return None;
    }
    let body = &marker[COMMAND_DONE_PREFIX.len()..marker.len() - 1];
    std::str::from_utf8(body).ok()?.parse().ok()
}

fn hex_encode(bytes: &[u8]) -> String {
    const HEX: &[u8; 16] = b"0123456789abcdef";
    let mut out = String::with_capacity(bytes.len() * 2);
    for &byte in bytes {
        out.push(HEX[(byte >> 4) as usize] as char);
        out.push(HEX[(byte & 0x0f) as usize] as char);
    }
    out
}

fn hex_decode(value: &str) -> Result<Vec<u8>> {
    let bytes = value.as_bytes();
    if bytes.len() % 2 != 0 {
        return Err(AppError::Msg("invalid request".to_string()));
    }
    let mut out = Vec::with_capacity(bytes.len() / 2);
    let mut index = 0;
    while index < bytes.len() {
        let hi = from_hex(bytes[index])?;
        let lo = from_hex(bytes[index + 1])?;
        out.push((hi << 4) | lo);
        index += 2;
    }
    Ok(out)
}

fn from_hex(byte: u8) -> Result<u8> {
    match byte {
        b'0'..=b'9' => Ok(byte - b'0'),
        b'a'..=b'f' => Ok(byte - b'a' + 10),
        b'A'..=b'F' => Ok(byte - b'A' + 10),
        _ => Err(AppError::Msg("invalid request".to_string())),
    }
}
